packages/host/app/components/matrix/register-user.gts

```diff
@@ -240,13 +435,195 @@ export default class RegisterUser extends Component<Args> {
     }
   }
 
+  @action private resendValidation() {
+    if (this.state.type === 'initial' || this.state.type === 'login') {
+      throw new Error(
+        `invalid state: cannot resendValidation() in state ${this.state.type}`,
+      );
+    }
+
+    this.state.sendAttempt++;
+    let { clientSecret, sendAttempt } = this.state;
+    this.validateEmail.perform(clientSecret, sendAttempt);
+  }
+
+  @action private doValidation() {
+    this.validateEmail.perform();
+  }
+
+  @action private cancel() {
+    localStorage.removeItem('email-validation');
+    this.args.onCancel();
+  }
+
+  private validateEmail = restartableTask(
+    async (clientSecret: string = uuidv4(), sendAttempt: number = 1) => {
+      if (!this.email) {
+        throw new Error(
+          `bug: should never get here: validate button disabled when no email`,
+        );
+      }
+      let email = this.email;
+      let res: IRequestTokenResponse | undefined;
+      try {
+        res = await this.matrixService.requestRegisterEmailToken(
+          email,
+          clientSecret,
+          sendAttempt,
+        );
+      } catch (e: any) {
+        if (e.status === 400 && e.data.errcode === 'M_THREEPID_IN_USE') {
+          this.emailError = e.message;
+          return;
+        }
+        throw e;
+      }
+      let { sid } = res;
+
+      // only need to kick off the check validation poll after
+      // the first vaidation attempt
+      if (sendAttempt === 1) {
```

Comment:
This isn’t necessarily true. If the user leaves the app and then comes back to it, we restore the email validation state from local storage. In which case they may have left the app after performing a resend. We’ll need to use new component state to know if validation poll is still running